### ISSUE TITLE
feat(executor): Hashjoin should stream to make best use of 

### ISSUE THREAD

**Issue by jon-chuang** [singularity-data/risingwave#2897]
With https://github.com/singularity-data/risingwave/issues/2428 we observe the issue that there is a need to fetch entries per update, but sometimes, updates come in batches of 1 or 2. Thus, we cannot exploit concurrency because the system is still written in a batch update fashion.

Hence, we should properly exploit concurrent streaming by allowing new messages to be added to a queue instantly, multiplexed into sync_queue and async_queue, sync_queue is a dequeue of work that can be performed instantly, async queue is an unbounded queue of work that completes async. Hopefully, we can reduce the overall latency this way.

**Comment by TennyZhuang** [comment]
> updates come in batches of 1 or 2.

How does it happen? Which executor produces the small chunk?

**Comment by fuyufjh** [comment]
> updates come in batches of 1 or 2.

Same question. Can this be solved by turning the checkpoint interval larger?

**Comment by lmatz** [comment]
I guess the point is that executors should not wait for an input chunk to be fully processed before processing the next input chunk.

To generalize a bit, we should do things as if vectorization provides 0 benefits. But since vectorization actually makes things faster, we should have a tradeoff/balance here.

**Comment by yuhao-su** [comment]
> To generalize a bit, we should do things as if vectorization provides 0 benefits. But since vectorization actually makes things faster, we should have a tradeoff/balance here.

Vectorization can be preserved in `sync_queue` here. For `async_queue` there might be a tradeoff.

**Comment by jon-chuang** [comment]
>  But since vectorization actually makes things faster, we should have a tradeoff/balance here.

I believe the relevant issue is https://github.com/singularity-data/risingwave/issues/2755 for getting the benefits of vectorization. I believe that improving the chunk size consistency is an orthogonal improvement. 

The tradeoff is latency v.s. throughput. So the solution I thought best was to wait for a fixed timeout (proposed to be 1ms) since last message before forwarding accumulated chunks. Every operator that does accumulation should follow this behaviour.

> How does it happen? Which executor produces the small chunk?

I believe it was tpch q14 and q3

> Vectorization can be preserved in sync_queue here. For async_queue there might be a tradeoff.

I don't think vectorization makes that much of a difference for stateful operators, which must spend time doing random lookup at best and at worst, async I/O. Correct me if I'm wrong. Probably vectorization is only useful for simple agg/simple arithmetic expressions (e.g. projection expression)?

**Comment by jon-chuang** [comment]
> Same question. Can this be solved by turning the checkpoint interval larger? 

How to do this?

**Comment by skyzh** [comment]
> > Same question. Can this be solved by turning the checkpoint interval larger?
> 
> How to do this?

In `risingwave.toml` we have a checkpoint interval config. Simply update it and start risedev cluster.

**Comment by jon-chuang** [comment]
I had the checkpoint interval set to 100ms. 

What I found was that it was the updates which would result in cache hit that were very few in number. I guess it was usually one side of the updates (the right side prior to ) that was producing very few rows/s.

Should be investigated further... (with metrics)

**Comment by yuhao-su** [comment]
Also might cause a problem producing duplicate pks... can be resolved if we make join key part of pk?
```
pk:       1st col
Join key: 2nd col

I 1 2 unmatched
D 1 2 unmatched
I 1 3 matched
D 1 3 matched

After reorder due to async fetching

I 1 3 matched
I 1 2 unmatched
D 1 2 unmatched
D 1 3 matched

This produce 2 identical insert PK to the downstream
```

**Comment by fuyufjh** [comment]
> I guess the point is that executors should not wait for an input chunk to be fully processed before processing the next input chunk, if CPU is idle.

Yes, I agree this is his point. But I think this proposal would not improve the latency directly, because latency is determined by barrier, and barrier cannot pass until all pending messages in `async_queue` to be processed. 

Instead, this might improve the throughput by batching all "cache-missed" data together to emit requests to S3.

**Comment by jon-chuang** [comment]
Actually some of the discussion from my end might be wrong, there is indeed vectorization for hash tables (vertical vectorization). It's basically fine to apply it in our case as we always need to first probe the hashmap cache prior to multiplexing into sync/async.

**Comment by jon-chuang** [comment]
> In another words, the quick messages do not help to improve the latency of barriers.

Actually, I disagree here.

Consider the following:
arrival time: 0ms message(10 sync, 10 async) async I/O: latency 5ms
arrival time: 1ms message(10 sync, 10 async) async I/O: latency 5ms

Total time for main: 5 * 10 * 2 = 100ms
Total time for concurrent fetch (https://github.com/singularity-data/risingwave/issues/2428): 5 + 5 = 10ms
Total time for this streaming messages (this issue): 5 + 1 = 6ms

Basically, even with concurrent fetch, if your message frequency < async I/O latency, you will expand out that latency to the async I/O latency. Rather than combatting with more parallelism (more shards), you could just process more messages' async I/O concurrently by adopting the streaming approach.

I also tend to think that the faster we allow the downstream operators to do useful work, the better utilization and latency we can get from our system.

**Comment by jon-chuang** [comment]
> Also might cause a problem producing duplicate pks... can be resolved if we make join key part of pk?

From our discussion offline @yuhao-su and I concluded this will cause unnecessary columns to be present in downstream output of joins, especially bad for many-way joins.

We discussed and I think I have a solution, I will write an RFC/design doc. Basically, its just having PK-dependent updates wait in special queues until their prior update is completed.

**Comment by fuyufjh** [comment]
> > In another words, the quick messages do not help to improve the latency of barriers.
> 
> 
> 
> Actually, I disagree here.
> 
> 
> 
> Consider the following:
> 
> arrival time: 0ms message(10 sync, 10 async) async I/O: latency 5ms
> 
> arrival time: 1ms message(10 sync, 10 async) async I/O: latency 5ms
> 
> 
> 
> Total time for main: 5 * 10 * 2 = 100ms
> 
> Total time for concurrent fetch (https://github.com/singularity-data/risingwave/issues/2428): 5 + 5 = 10ms
> 
> Total time for this streaming messages (this issue): 5 + 1 = 6ms
> 
> 
> 
> Basically, even with concurrent fetch, if your message frequency < async I/O latency (due to high streaming throughput), you will expand out that latency to the async I/O latency. Rather than combatting with more parallelism (more shards), you could just process more messages' async I/O concurrently by adopting the streaming approach.
> 
> 
> 
> I also tend to think that the faster we allow the downstream operators to do useful work, the better utilization and latency we can get from our system. For instance, downstream async I/O could begin earlier as a result.
> 
> 
> 
> Probably there is some math to show this by modelling as fluid flows...

OK, I misunderstood it before. This approach seems an enhanced version of concurrent fetch to me.

Furthermore, I think it's possible to preserve the message order in this case. Similar to the TCP protocol, a later packet/fetch may be received first, but we can just put it in the queue until all its prior packet done

**Comment by jon-chuang** [comment]
>  a later packet/fetch may be received first, but we can just put it in the queue until all its prior packet done

Yes, but actually, all we need to preserve is message ordering for a given PK. Would you agree?

**Comment by fuyufjh** [comment]
> > a later packet/fetch may be received first, but we can just put it in the queue until all its prior packet done
> 
> Yes, but actually, all we need to preserve is update ordering for a given PK. Would you agree?

In another word, only `insert` messages (without any delete/update event between them) are reorderable?

**Comment by jon-chuang** [comment]
> In another word, only `insert` messages (without any delete/update event between them) are reorderable?

Actually, no, invidivual insert, delete, update can be reordered as long as the order of the row + op is preserved for a given PK.

**Comment by fuyufjh** [comment]
> > In another word, only `insert` messages (without any delete/update event between them) are reorderable?
> 
> Actually, no, invidivual insert, delete, update can be reordered as long as the order of the row + op is preserved for a given PK. That is unless PK can change for a given row...

Yes, but this algorithm need to maintain a set of PK in this epoch, which seems a little complex to me. 

In practice, I prefer the simplest solution - just preserve the order of all rows, no matter what they are. Will this harm the performance?

**Comment by jon-chuang** [comment]
well, it would if majority of the rows hit cache, but a small number result in async I/O. Then there would be a large fraction of rows waiting for rows ahead of them. 

I agree however that that's a simpler solution.

**Comment by jon-chuang** [comment]
> That is unless PK can change for a given row...

Actually, is this even possible? Seems like no right? An update only occurs if a row identified by PK changes.

**Comment by jon-chuang** [comment]
> Yes, but this algorithm need to maintain a set of PK in this epoch, which seems a little complex to me.

Its slightly complex, but maybe I will write up a solution and we can judge if its worth it to test out. My sense is that its not significantly complex.

**Comment by lmatz** [comment]
> > Yes, but this algorithm need to maintain a set of PK in this epoch, which seems a little complex to me.
> 
> Its slightly complex, but maybe I will write up a solution and we can judge if its worth it to test out. My sense is that its not significantly complex.

Is it possible to implement it in a way that we still can disallow the reordering of rows but only do concurrent state fetching for consecutive data chunks? 

I think it would be very helpful for us to understand what each technique can bring to us 🤔

**Comment by jon-chuang** [comment]
> Is it possible to implement it in a way that we still can disallow the reordering of rows but only do concurrent state fetching for consecutive data chunks?

Yes. But as discussed:

> well, it would if majority of the rows hit cache, but a small number result in async I/O. Then there would be a large fraction of rows waiting for rows ahead of them.

Basically, if you expect ocassional cache miss but high cache hit, downstream can process new rows much more quickly.

---

Also, here is the pseudocode for hashjoin (its not that complex! just an additional small hashmap for pending PK entries):

```rust
fn process_chunk(&mut self, chunk: StreamChunk) {
    for row in chunk {
      // delay the processing until pk dependency completes
      if let Some(mut in_progress_rows) = self.pk_in_progress.get_mut(row.pk) { 
        // clone it so we can throw away this chunk
        in_progress_rows.push(pk.row.to_owned_row());
      } else {
        // We are ok to process the row in any order because there is no outstanding
        // task for that PK.
        self.process_rows_out_of_order(row)
    }
}
    
fn process_rows_out_of_order(&mut self, row: &RowRef) {
  if let Some(entries) = self.cache.get_both_sides(row.join_key) {
    // if no dependencies and available in cache, process instantly
    self.process(row, entries);
  } else {
    // else, not available locally, add to async queue
    let entry = self.async_queue.entry(row.join_key).or_insert_with(Vec::new);
    if entry.is_empty() {
        // trigger the fetch of the join key data
        self.task_stream.push(async move {
            let jk_entries = self.cache.fetch_both_sides(row.join_key).await;
            
            // once we fetch the JK entry via async I/O, we can process it 
            // and queue any dependent entries.
            for row in self.async_queue.get(row.join_key) {
                self.process(row, entries);
                let pk_entry = self.pk_in_progress.get_mut(row.pk).unwrap();
                // remove this row from pending
                pk_entry.pop_head();
                // process the next row for this PK
                if !pk_entry.is_empty() {
                    self.process_row_out_of_order(pk_entry[0]);
                }
            }
        });
    }
    // mark this row as pending for its PK.
    entry.push(pk.row.to_owned_row());
  }
}
```

**Comment by jon-chuang** [comment]
Note that flink also has a knob allowing trading off latency for throughput in terms of the buffering timeout:

![](https://www.ververica.com/hs-fs/hubfs/Imported_Blog_Media/buffer_timeout-1.png?width=900&name=buffer_timeout-1.png)
https://www.ververica.com/blog/high-throughput-low-latency-and-exactly-once-stream-processing-with-apache-flink
